src/main/java/org/aksw/simba/lemming/creation/Inferer.java

```diff
@@ -301,67 +316,67 @@ public OntModel readOntology(String filePath, String base) {
-			boolean isSame = false;
-			
-			if (curURI!=null && !uriNodeMap.containsKey(curURI)) {
-				List eqsList = null;
+
+			if (curURI!=null) {
+
+				//find equivalent classes if possible
+				List<T> eqsList = null;
 				try {
 					if (currentResource.isProperty())
-						eqsList = currentResource.asProperty().listEquivalentProperties().toList();
+						eqsList = (List<T>) currentResource.asProperty().listEquivalentProperties().toList();
 					if (currentResource.isClass())
-						eqsList = currentResource.asClass().listEquivalentClasses().toList();
+						eqsList = (List<T>) currentResource.asClass().listEquivalentClasses().toList();
 				} catch (ConversionException e) {
 					LOGGER.warn(
 							"Cannot convert the equivalents. The ontology does not have any further info on the equivalents of {}.",
 							currentResource.toString());
 				}
 
-				if (eqsList != null && !eqsList.isEmpty()) {
-					stack.addAll(eqsList);
-				}
-				
 				//node to where we want to add the info to
 				Equivalent curNode = null;
-
-				// check to which node do we need to add this info to
-				Iterator<Equivalent> propIterator = elements.iterator();
-				while (propIterator.hasNext()) {
-					curNode = propIterator.next();
-					isSame = curNode.containsElement(currentResource);
-					if (isSame) {
-						curNode.addEquivalent(currentResource);
-						break;
+				if(eqsList == null || eqsList.isEmpty()){
+					if(!uriNodeMap.containsKey(curURI)){
+						curNode = new Equivalent(currentResource);
+						uriNodeMap.put(curURI, curNode);
 					}
-				}
+				}else{
 
-				// if not, create new one
-				if (!isSame) {
-					curNode = new Equivalent(currentResource);
-					elements.add(curNode);
-				}
-				
-				// add the node to the map with the URI and add the equivalents (if existing) to the node object
-				if (curNode != null) {
-					uriNodeMap.put(curURI, curNode);
-					if(eqsList != null) {
-						curNode.addEquivalentGroup((Set) eqsList.stream().collect(Collectors.toSet()));
+					if(uriNodeMap.containsKey(curURI)){
+						curNode = uriNodeMap.get(curURI);
+					}
+					Map<T, Equivalent> localMap = new HashMap<>();
+					for(T re : eqsList){
+						if(re.getURI()!=null && uriNodeMap.containsKey(re.getURI())){
+							localMap.put(re, uriNodeMap.get(re.getURI()));
+						}
+					}
+					if(!localMap.isEmpty()){
+						for(T re : localMap.keySet()){
+							if(curNode == null){
+								curNode = localMap.get(re);
+							}else if (curNode != localMap.get(re)){
+								curNode.addEquivalentGroup(localMap.get(re).getEquiResources());
+							}
+						}
+					}
+					if(curNode==null){
+						curNode = new Equivalent(currentResource);
+					}
+					curNode.addEquivalent(currentResource);
+					curNode.addEquivalentGroup(eqsList.stream().collect(Collectors.toSet()));
+					for(Object s : curNode.getEquivalents()){
+						uriNodeMap.put((String) s, curNode);
 					}
 				}
 			}
 		}
 		return uriNodeMap;
 	}

```

Comment:
I wonder if this method cannot be further optimized. 
I think some OWL Jena Reasoner can infer all the missing equivalentProperty and equivalentClass statements from the ontology. See  [Jena OWL Inference](https://jena.apache.org/documentation/inference/#OWLcoverage). However, the number of vertices of each dataset's target graph is likely to change with a different reasoner and would need to be updated.

If we change the reasoner with which the ontology is read with from the default one to ` ModelFactory.createOntologyModel(OntModelSpec.OWL_LITE_MEM_RULES_INF);`  in the dataset managers. It would simplify this searchEquivalents method, since we would only need to:
```
for (T currentResource : ontElements) {
  //  list equivalent resources of currentResource 
  //  create an Equivalent object, add the resources obtained in the previous step and update uriNodeMap 
}
```
@MichaelRoeder what do you think?

Reply:
It would be nice if we could simply use an existing implementation. However, my experience with the Jena implementations are not good. They follow very strictly the theoretical rules and add a lot of information that we are not interested in. Hence, they will be slower and increase the size of our ontology a lot.

However, I agree that the method can be optimized further. Apart from smaller improvements (e.g., line 371 is a very costly transformation of a list to a set), I am also afraid that it could contain a severe bug. If I understand the code correctly (it is not so easy to understand the work flow of the method so I am not 100% sure :sweat_smile: ), line 373 could override information that was already in the map but will be lost since the `Equivalent` instances of all other URIs are overridden with the information that we get from the current URI. This does not take into account that there might have been information added before. :thinking: 

First, I would like to clarify whether we really need the class `Equivalent`. To me, it seems to make things more complicated and it also seems to consume more space than necessary :thinking: 
I saw that it also overrides domain and range information of properties. I think this isn't something that such a class should do. Instead, the Inferer should do that in a separate method (after it identified the properties that are equivalent to each other).

Wouldn't a `Map<T, Set<T>>` be sufficient to store equivalents?

Reply:
The `Equivalent` class was created to store the equivalents, to store an "uniform" URI that would be used to replace the equivalent classes/properties in statements and to keep track of the domain and range information of all equivalent properties. 

But yes, if we pass these class' responsibilities to the Inferer, then it should also be enough to have a `Map<T, Set<T>>` to store the equivalents.

Reply:
I see. Well, in that case, I would suggest to remove the class and separate these steps, i.e.,
1. Search for equivalent IRIs with `Map<T, Set<T>>`
2. For each set of equivalent IRIs, choose an IRI that represents it. (for properties, ensure that domain and range are adapted)

The second step means that classes have to be handled first right? Otherwise we end up with classes in our domain and range definition, that may have been replaced with another class :thinking:

Reply:
Done